File: 004.Median-of-Two-Sorted-Arrays/4-Median-of-Two-Sorted-Arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) 
    {
        int m=nums1.size();
        int n=nums2.size();
        
        if ((m+n)%2==1)
            return FindKthSmallest(nums1,0,m,nums2,0,n,(m+n+1)/2);
        else
            return (FindKthSmallest(nums1,0,m,nums2,0,n,(m+n)/2)+FindKthSmallest(nums1,0,m,nums2,0,n,(m+n)/2+1))/2;
        
    }
    
    double FindKthSmallest(vector<int>& nums1, int a, int m, vector<int>& nums2, int b, int n, int k)
    {
        //cout<<a<<" "<<m<<" "<<b<<" "<<n<<endl;
        
        if (m>n) return FindKthSmallest(nums2,b,n,nums1,a,m,k);
        
        if (m==0) return nums2[b+k-1];
        
        if (k==1) return min(nums1[a],nums2[b]);
        
        int k1=min(m,k/2);
        int k2=k-k1;
        
        if (nums1[a+k1-1]<nums2[b+k2-1])
            return FindKthSmallest(nums1,a+k1,m-k1,nums2,b,n,k-k1);
        else
            return FindKthSmallest(nums1,a,n,nums2,b+k2,n-k2,k-k2);
    }
};

File: Binary_Search/004.Median-of-Two-Sorted-Arrays/4. Median of Two Sorted Arrays.py
def findMedianSortedArrays(nums1, nums2):
    n1, n2 = len(nums1), len(nums2)
    if n1 > n2:
        return findMedianSortedArrays(nums2, nums1) #cambian si es mas grande de matrices

    n = n1 + n2 #el total de todos los numeros

    left = (n1 + n2 + 1) // 2  #longi de la izquierda mitad
    #de aqui abajo aplicamos busqueda binaria
    low, high = 0, n1

    while low <= high:
        mid1 = (low + high) // 2
        mid2 = left - mid1

        #aqui calculamos los l y r que nos confundio en lo escrito

        l1, l2, r1, r2 = float('-inf'), float('-inf'), float('inf'), float('inf')

        if mid1 < n1:
            r1 = nums1[mid1]
        if mid2 < n2:
            r2 = nums2[mid2]
        if mid1 - 1 >= 0:
            l1 = nums1[mid1 - 1]
        if mid2 - 1 >= 0:
            l2 = nums2[mid2 - 1]

        if l1 <= r2 and l2 <= r1:
            if n % 2 == 1:
                return max(l1, l2)
            else:
                return (float(max(l1, l2)) + float(min(r1, r2))) / 2.0

        elif l1 > l2:
            high = mid1 - 1
        else:
            low = mid1 + 1
    return 0
